trie/stacktrie.go

```diff
@@ -0,0 +1,569 @@
+// Copyright 2020 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package trie
+
+import (
+	"io"
+	"bytes"
+
+	"github.com/ethereum/go-ethereum/common"
+	"golang.org/x/crypto/sha3"
+)
+
+// StackTrie is a trie implementation that expects keys to be inserted
+// in order. Once it determines that a subtree will no longer be inserted
+// into, it will hash it and free up the memory it uses.
+type StackTrie struct {
+	nodeType  uint8          // node type (as in branch, ext, leaf)
+	val       []byte         // value contained by this node if it's a leaf
+	key       []byte         // key chunk covered by this (full|ext) node
+	keyOffset int            // offset of the key chunk inside a full key
+	children  [16]*StackTrie // list of children (for fullnodes and exts)
+
+	db *Database // Pointer to the commit db, can be nil
+}
+
+// NewStackTrie allocates and initializes an empty trie.
+func NewStackTrie() *StackTrie {
+	return &StackTrie{
+		nodeType: emptyNode,
+	}
+}
+
+func newLeaf(ko int, key, val []byte) *StackTrie {
```

Comment:
Please rephrase what you mean. The offset is from the start of the inserted key and is correct as far as I can tell.

Reply:
For example, https://github.com/gballet/go-ethereum/blob/simple-pruner-with-stacktrie/trie/stacktrie.go#L223 here the passed offset is not the real offset from the start of the inserted key.

Reply:
I see. Well, it can't be from the start of the inserted key, because `st.key` has been inserted before and then sliced up to reach that point, so it would be possible to re-slice the key (we know the real offset). Or keep it entirely and use the offset. Ok, I think I see what you mean. Let me think it over.